tests/unit/lib/intrinsic_resolver/test_intrinsic_resolver.py

```diff
@@ -0,0 +1,1443 @@
+        intrinsic = {
+            "Fn::Join": [",", ["a", "b", "c", "d"]]
+        }
+        result = self.resolver.intrinsic_property_resolver(intrinsic)
+        self.assertEquals(result, "a,b,c,d")
+
+    def test_nested_fn_join(self):
+        intrinsic_base_1 = {
+            "Fn::Join": [",", ["a", "b", "c", "d"]]
+        }
+        intrinsic_base_2 = {
+            "Fn::Join": [";", ["g", "h", "i", intrinsic_base_1]]
+        }
+        intrinsic = {
+            "Fn::Join": [":", [intrinsic_base_1, "e", "f", intrinsic_base_2]]
+        }
+        result = self.resolver.intrinsic_property_resolver(intrinsic)
+        self.assertEquals(result, "a,b,c,d:e:f:g;h;i;a,b,c,d")
+
+    def test_delimiter_fn_join(self):
+        intrinsic_base_1 = {
+            "Fn::Join": ["", ["t", "e", "s", "t"]]
+        }
+        intrinsic = {
+            "Fn::Join": [intrinsic_base_1, ["a", "e", "f", "d"]]
+        }
+        result = self.resolver.intrinsic_property_resolver(intrinsic)
+        self.assertEquals(result, "atestetestftestd")
+
+    @parameterized.expand([
+        ("Fn::Join should fail for values that are not lists: {}".format(item), item)
+        for item in
+        [True, False, "Test", {}, 42, None]
+    ])
+    def test_fn_join_arguments_invalid_formats(self, name, intrinsic):
+        with self.assertRaises(InvalidIntrinsicException, msg=name):
+            self.resolver.intrinsic_property_resolver({
+                "Fn::Join": intrinsic
+            })
+
+    @parameterized.expand([
+        ("Fn::Join should fail if the first argument does not resolve to a string: {}".format(item), item)
+        for item in
+        [True, False, {}, 42, None]
+    ])
+    def test_fn_join_delimiter_invalid_type(self, name, delimiter):
+        with self.assertRaises(InvalidIntrinsicException, msg=name):
+            self.resolver.intrinsic_property_resolver({
+                "Fn::Join": [delimiter, []]
+            })
+
+    @parameterized.expand([
+        ("Fn::Join should fail if the list_of_objects is not a valid list: {}".format(item), item)
+        for item in
+        [True, False, {}, 42, "t", None]
+    ])
+    def test_fn_list_of_objects_invalid_type(self, name, list_of_objects):
+        with self.assertRaises(InvalidIntrinsicException, msg=name):
+            self.resolver.intrinsic_property_resolver({
+                "Fn::Join": ["", list_of_objects]
+            })
+
+    @parameterized.expand([
+        ("Fn::Join should require that all items in the list_of_objects resolve to string: {}".format(item),
+         item)
+        for item in
+        [True, False, {}, 42, None]
+    ])
+    def test_fn_join_items_all_str(self, name, single_obj):
+        with self.assertRaises(InvalidIntrinsicException, msg=name):
+            self.resolver.intrinsic_property_resolver({
+                "Fn::Join": ["", ["test", single_obj, "abcd"]]
+            })
+
+
+class TestIntrinsicFnSplitResolver(TestCase):
```

Comment:
This goes for all of them, we should have some negative case tests too. Like what happens when you split but the delimiter is not in the string.